demos/free_surface/free_surface.py

```diff
@@ -0,0 +1,355 @@
+    time += dt
+
+    # Solve Stokes sytem:
+    stokes_solver.solve()
+
+    # Temperature system:
+    energy_solver.solve()
+
+    # Compute diagnostics:
+    energy_conservation = abs(abs(gd.Nu_top()) - abs(gd.Nu_bottom()))
+
+    # Calculate L2-norm of change in temperature:
+    maxchange = sqrt(assemble((T - energy_solver.T_old) ** 2 * dx))
+
+    # Log diagnostics:
+    plog.log_str(
+        f"{timestep} {time} {float(delta_t)} {maxchange} "
+        f"{gd.u_rms()} {gd.u_rms_top()} {gd.ux_max(top_id)} {gd.Nu_top()} "
+        f"{gd.Nu_bottom()} {energy_conservation} {gd.T_avg()} "
+        f"{z.subfunctions[2].dat.data.min()} {z.subfunctions[2].dat.data.max()}"
+    )
+
+    # Leave if steady-state has been achieved:
+    if maxchange < steady_state_tolerance:
+        log("Steady-state achieved -- exiting time-step loop")
+        break
+
+# At the end of the simulation, once a steady-state has been achieved, we close our logging file
+# and checkpoint steady state temperature and Stokes solution fields to disk. These can later be
+# used to restart a simulation, if required.
+
+# +
+plog.close()
+
+with CheckpointFile("Final_State.h5", "w") as final_checkpoint:
+    final_checkpoint.save_mesh(mesh)
+    final_checkpoint.save_function(T, name="Temperature")
+    final_checkpoint.save_function(z, name="Stokes")
+# -
+
+# Let's use the python package **PyVista** to plot the temperature field through time. We will use the calculated free surface height to artifically scale the mesh in the vertical direction. We have exaggerated the vertical stretching by a factor of 1.5, **BUT...** it is important to remember this is just for ease of visualisation - the mesh is not moving in reality!
+
+# +
+# Read the PVD file
+reader = pv.get_reader('output.pvd')
+data = reader.read()[0]  # MultiBlock mesh with only 1 block
+
+# Create a plotter object
+plotter = pv.Plotter(shape=(1, 1), border=False, notebook=True, off_screen=False)
+
+# Open a gif
+plotter.open_gif("temperature_warp.gif")
+
+# Make a colour map
+boring_cmap = plt.get_cmap("coolwarm", 25)
+
+for i in range(len(reader.time_values)):
+    reader.set_active_time_point(i)
+    data = reader.read()[0]
+
+    # Artificially warp the output data in the vertical direction by the free surface height
+    warped = data.warp_by_scalar(scalars='eta', normal=(0, 1, 0), factor=1.5)
+
+    # Add the warped temperature field to the frame
+    plotter.add_mesh(
+        warped,
+        scalars="Temperature",
+        lighting=False,
+        show_edges=False,
+        clim=[0, 1],
+        cmap=boring_cmap,
+        scalar_bar_args={'position_x': 0.2, 'position_y': 0.05}
+    )
+    arrows = data.glyph(orient="Velocity", scale="Velocity", factor=0.001, tolerance=0.05)
+    plotter.add_mesh(arrows, color="black")
+    # Centre camera on xy plane and write frame
+    plotter.camera_position = "xy"
+    plotter.write_frame()
+    plotter.clear()
+
+# Closes and finalizes movie
+plotter.close()
+
+
+# -
```

Comment:
Should this section have the `"active-ipynb"` tag? It is currently generating a CI execution error, as `pyvista` and `matplotlib` are not imported.

Reply:
Thanks for picking this up @Patol75 also as you mentioned offline I've added imageio to the demos dependency list for gadopt in the pytoml.